src/externalized/strategic/UndergroundSectorModel.cc

```diff
@@ -0,0 +1,126 @@
+
+	return u;
+}
+
+std::array<uint8_t, NUM_DIF_LEVELS> readIntArray(const rapidjson::Value& obj, const char* fieldName)
+{
+	std::array<uint8_t, NUM_DIF_LEVELS> vals = {};
+	if (!obj.HasMember(fieldName))
+	{
+		return vals;
+	}
+	auto arr = obj[fieldName].GetArray();
+	if (arr.Size() != NUM_DIF_LEVELS)
+	{
+		SLOGE(ST::format("The number of values in {} is not same as NUM_DIF_LEVELS({})", fieldName, NUM_DIF_LEVELS));
+		throw std::runtime_error("The number of elements does not match the number of difficulty levels");
+	}
+	for (auto i = 0; i < NUM_DIF_LEVELS; i++)
+	{
+		vals[i] = static_cast<uint8_t>(arr[i].GetUint());
+	}
+
+	return vals;
+}
+
+UndergroundSectorModel* UndergroundSectorModel::deserialize(const rapidjson::Value& obj)
+{
+	auto sectorString = obj["sector"].GetString();
+	if (!IS_VALID_SECTOR_SHORT_STRING(sectorString))
+	{
+		SLOGE(ST::format("'{}' is not valid sector", sectorString));
+		throw std::runtime_error("Not a valid sector");
+	}
+	uint8_t sectorId = SECTOR_FROM_SECTOR_SHORT_STRING(sectorString);
+	uint8_t sectorZ = obj["sectorLevel"].GetUint();
+	if (sectorZ == 0 || sectorZ > 3)
+	{
+		SLOGE("Sector level must be between 1 and 3");
+		throw std::runtime_error("Sector level is invalid");
+	}
+
+	uint8_t adjacencyFlag = NO_ADJACENT_SECTOR;
+	if (obj.HasMember("adjacentSectors") && obj["adjacentSectors"].IsArray())
+	{
+		auto adjacencies = obj["adjacentSectors"].GetArray();
+		for (auto& el : adjacencies)
+		{
+			const char* adj = el.GetString();
+			if (strlen(adj) != 1)
+			{
+				SLOGE(ST::format("'{}' is not a valid adjacency direction.", adj));
+				throw std::runtime_error("Not a valid adjacency direction");
+			}
+			switch (adj[0])
+			{
+			case DIRECTION_NORTH:
+				adjacencyFlag |= NORTH_ADJACENT_SECTOR;
+				break;
+			case DIRECTION_EAST:
+				adjacencyFlag |= EAST_ADJACENT_SECTOR;
+				break;
+			case DIRECTION_SOUTH:
+				adjacencyFlag |= SOUTH_ADJACENT_SECTOR;
+				break;
+			case DIRECTION_WEST:
+				adjacencyFlag |= WEST_ADJACENT_SECTOR;
+				break;
+			default:
+				SLOGE(ST::format("{} is not a valid direction.", adj[0]));
+				throw std::runtime_error("Invalid direction");
+			}
+		}
+	}
+	
+	return new UndergroundSectorModel(
+		sectorId, sectorZ, adjacencyFlag,
+		readIntArray(obj, "numTroops"), readIntArray(obj, "numElites"), readIntArray(obj, "numCreatures"),
+		readIntArray(obj, "numTroopsVariance"), readIntArray(obj, "numElitesVariance"), readIntArray(obj, "numCreaturesVariance")
+	);
+}
+
+void UndergroundSectorModel::validateData(const std::vector<const UndergroundSectorModel*> ugSectors)
+{
+	// check for existence of hard-coded references
+	// the list below is based on the occurrences of FindUnderGroundSector in codebase
+	std::array<uint8_t, 3> basementSectors[] = {
+		{ 13, 5,  1 },  // Draassen mine
+		{ 9,  8,  1 },  // Cambria mine
+		{ 14, 10, 1 },  // Alma mine
+		{ 4,  8,  2 }   // Grumm mine
```

Comment:
So basically modders can only add underground sectors, not move or remove them. I guess this is not a problem for Wildfire, since the maps are so similar?

Reply:
Yes for now, and only these 4 sectors, because  of creature spreading:

https://github.com/ja2-stracciatella/ja2-stracciatella/blob/2ad95c3a7b26ca6c8436e3389173b3be8c9de80a/src/game/Strategic/Creature_Spreading.cc#L332-L333

Alternatively, we can add null-check in Creature_Spreading.cc, and we only WARN on missing.

And later when we externalize creature spreading or make it optional,  we can remove this check.

Reply:
It's confusing, since in the json all the mine sectors are in the required part, not just these entrances. But yeah, should also be externalized eventually.

Reply:
Oh right. The line off comment "Do all of the mandatory underground mine sectors" was copied from the original code.

Those other mine sectors can be moved and removed, as long as they are consistent with Data/Maps. Only these 4 creature lair entrances are referenced by code.

Reply:
Removed that line of confusing comment.